instructor/process_response.py

```diff
@@ -443,117 +514,114 @@ def handle_response_model(
-        elif mode == Mode.VERTEXAI_TOOLS:
-            from instructor.client_vertexai import vertexai_process_response
-
-            contents, tools, tool_config = vertexai_process_response(
-                new_kwargs, response_model
-            )
-
-            new_kwargs["contents"] = contents
-            new_kwargs["tools"] = tools
-            new_kwargs["tool_config"] = tool_config
-        elif mode == Mode.VERTEXAI_JSON:
-            from instructor.client_vertexai import vertexai_process_json_response
-
-            contents, generation_config = vertexai_process_json_response(
-                new_kwargs, response_model
-            )
-
-            new_kwargs["contents"] = contents
-            new_kwargs["generation_config"] = generation_config
-        else:
-            raise ValueError(f"Invalid patch mode: {mode}")
-
-    # Handle Cohere Response Model Case
-    elif response_model is None and mode in {
-        Mode.COHERE_JSON_SCHEMA,
-        Mode.COHERE_TOOLS,
-    }:
-        messages = new_kwargs.pop("messages", [])
-        chat_history = []
-        for message in messages[:-1]:
-            # format in Cohere's ChatMessage format
-            chat_history.append(
-                {
-                    "role": message["role"],
-                    "message": message["content"],
-                }
-            )
-        new_kwargs["message"] = messages[-1]["content"]
-
-        new_kwargs["chat_history"] = chat_history
-        if "model_name" in new_kwargs and "model" not in new_kwargs:
-            new_kwargs["model"] = new_kwargs.pop("model_name")
-        new_kwargs.pop("strict", None)
+    return response_model
+
+
+def handle_response_model(
+    response_model: type[T] | None, mode: Mode = Mode.TOOLS, **kwargs: Any
+) -> tuple[type[T] | None, dict[str, Any]]:
+    """
+    Handles the response model based on the specified mode and prepares the kwargs for the API call.
+
+    Args:
+        response_model (type[T] | None): The response model to be used for parsing the API response.
+        mode (Mode): The mode to use for handling the response model. Defaults to Mode.TOOLS.
+        **kwargs: Additional keyword arguments to be passed to the API call.
+
+    Returns:
+        tuple[type[T] | None, dict[str, Any]]: A tuple containing the processed response model and the updated kwargs.
+
+    This function prepares the response model and modifies the kwargs based on the specified mode.
+    It handles various modes like TOOLS, JSON, FUNCTIONS, etc., and applies the appropriate
+    transformations to the response model and kwargs.
+    """
+
+    new_kwargs = kwargs.copy()
+
+    if response_model is None:
+        if mode in {Mode.COHERE_JSON_SCHEMA, Mode.COHERE_TOOLS}:
+            # This is cause cohere uses 'message' and 'chat_history' instead of 'messages'
+            return handle_cohere_modes(new_kwargs)
+        return None, new_kwargs
+
+    if mode in {Mode.PARALLEL_TOOLS}:
```

Comment:
In the original mode code, we handled parallel tools differently too so I added it as a special case

```python
	if is_simple_type(response_model):
            response_model = ModelAdapter[response_model]

        if is_typed_dict(response_model):
            response_model: BaseModel = create_model(
                response_model.__name__,
                **{k: (v, ...) for k, v in response_model.__annotations__.items()},
            )

        # This a special case for parallel tools
        if mode == Mode.PARALLEL_TOOLS:
            assert (
                new_kwargs.get("stream", False) is False
            ), "stream=True is not supported when using PARALLEL_TOOLS mode"
            new_kwargs["tools"] = handle_parallel_model(response_model)
            new_kwargs["tool_choice"] = "auto"

            # This is a special case for parallel models
            response_model = ParallelModel(typehint=response_model)
            return response_model, new_kwargs
```

@jxnl